views/synergy-demo-view/src/components/SynergyDemoView/SynergyDemoView.tsx

```diff
@@ -1,81 +1,278 @@
+  }
+
+  async function findEntryType(itemId) {
+    const entryTypeLink = await perspective.get(
+      new LinkQuery({ source: itemId, predicate: "flux://entry_type" })
+    );
+    if (!entryTypeLink[0]) return "Task";
+    else return entryTypeToSubjectClass(entryTypeLink[0].data.target);
+  }
+
+  async function findChannel(itemId, channels) {
+    const parentLinks = await perspective.get(
+      new LinkQuery({ predicate: "ad4m://has_child", target: itemId })
+    );
+    return (
+      channels.find((c) => parentLinks.find((p) => p.data.source === c.id)) || {
+        id: "unlinked",
+      }
+    );
+  }
+
+  async function findEmbeddingMatches(
+    itemId: string,
+    allowedTypes: string[],
+    channels: any[]
+  ): Promise<any[]> {
+    // searches for items in the neighbourhood that match the search filters & have similar embedding scores
+    return await new Promise(async (resolveMatches: any) => {
+      // grab all embedding relationships
+      const allEmbeddingRelationships = (await SemanticRelationship.all(perspective)).filter(
+        (r) => !r.relevance
+      );
+      // find the source embedding
+      const sourceEmbeddingRelationship = allEmbeddingRelationships.find(
+        (r) => r.expression === itemId
+      );
+      const sourceEmbeddingEntity = new Embedding(perspective, sourceEmbeddingRelationship.tag);
+      const sourceEmbedding = await sourceEmbeddingEntity.get();
+      // loop through others & apply search filters
+      const embeddingRelationships = allEmbeddingRelationships.filter(
+        (r) => r.expression !== itemId
+      );
+      const otherValidEmbeddings = await Promise.all(
+        embeddingRelationships.map(async (relationship: any) => {
+          const { expression, tag } = relationship;
+          // if it doesn't match the search filters return null
+          const channel = await findChannel(expression, channels);
+          const type = await findEntryType(expression);
+          const wrongChannel = !filterSettings.includeChannel && channel.id === source;
+          const wrongType = !allowedTypes.includes(type);
+          if (wrongChannel || wrongType) return null;
+          // otherwise return the matching expression with its embedding
+          const embeddingEntity = new Embedding(perspective, tag);
+          const embedding = await embeddingEntity.get();
+          return {
+            baseExpression: expression,
+            channel,
+            type,
+            embedding: JSON.parse(embedding.embedding),
+          };
+        })
+      );
+      // generate similarity score for items
+      const embeddingWorker = new EmbeddingWorker();
+      embeddingWorker.onmessage = async (message) => {
+        resolveMatches(message.data);
+        worker.current?.terminate();
+      };
+      embeddingWorker.postMessage({
+        items: otherValidEmbeddings.filter((i) => i !== null),
+        sourceEmbedding: JSON.parse(sourceEmbedding.embedding),
+      });
```

Comment:
This is where we could use `cos_sim()` directly instead of using a worker thread